File: src/eko-core/prompt/builders/prompt-builder.ts
/**
 * 提示词构建器
 * 负责组装各个组件生成完整的提示词
 */

import { ComponentContext, PromptComponent, PromptBuildOptions, PromptVariant } from '../components/types'
import { getComponentRegistry } from '../components/registry'
import { EkoTemplateEngine } from '../template-engine'

/**
 * 提示词构建器类
 */
export class PromptBuilder {
  private templateEngine: EkoTemplateEngine

  constructor() {
    this.templateEngine = EkoTemplateEngine.getInstance()
  }

  /**
   * 构建提示词
   * @param context 组件上下文
   * @param options 构建选项
   */
  async build(context: ComponentContext, options: PromptBuildOptions = {}): Promise<string> {
    const registry = getComponentRegistry()
    await registry.load()

    const {
      components = [],
      templateOverrides = {} as Partial<Record<PromptComponent, string>>,
      skipMissing = true,
    } = options

    // 验证依赖关系
    const errors = registry.validateDependencies(components)
    if (errors.length > 0 && !skipMissing) {
      throw new Error(`组件依赖验证失败: ${errors.join(', ')}`)
    }

    // 按依赖关系排序组件
    const sortedComponents = registry.sortComponentsByDependencies(components)

    // 构建各个组件
    const componentResults: Record<string, string> = {}

    for (const componentId of sortedComponents) {
      const component = registry.getComponent(componentId)
      if (!component) {
        if (!skipMissing) {
          throw new Error(`组件不存在: ${componentId}`)
        }
        continue
      }

      try {
        // 创建组件副本以避免修改共享对象
        const componentCopy = { ...component }
        if (templateOverrides[componentId]) {
          componentCopy.template = templateOverrides[componentId]
        }

        // 创建临时上下文，包含覆盖的模板
        const contextWithTemplate = {
          ...context,
          _templateOverride: componentCopy.template,
        }

        const result = await component.fn(contextWithTemplate)

        if (result !== undefined) {
          componentResults[componentId] = result
        }
      } catch (error) {
        if (!skipMissing) {
          throw new Error(`组件 ${componentId} 构建失败: ${error}`)
        }
        console.warn(`组件 ${componentId} 构建失败，已跳过:`, error)
      }
    }

    // 组装最终提示词
    return this.assemblePrompt(componentResults)
  }

  /**
   * 使用预定义变体构建提示词
   * @param variant 提示词变体
   * @param context 组件上下文
   * @param additionalContext 额外上下文
   */
  async buildFromVariant(
    variant: PromptVariant,
    context: ComponentContext,
    additionalContext: Record<string, unknown> = {}
  ): Promise<string> {
    // 构建组件
    const options: PromptBuildOptions = {
      components: variant.components,
      additionalContext: { ...variant.defaultContext, ...additionalContext },
    }

    const componentResults = await this.buildComponents(context, options)

    // 使用变体模板组装
    const allContext = {
      ...componentResults,
      ...additionalContext,
      ...variant.defaultContext,
    }

    return this.templateEngine.resolve(variant.template, allContext)
  }

  /**
   * 构建组件内容
   */
  private async buildComponents(
    context: ComponentContext,
    options: PromptBuildOptions
  ): Promise<Record<string, string>> {
    const registry = getComponentRegistry()
    await registry.load()

    const { components = [], templateOverrides = {}, skipMissing = true } = options

    const sortedComponents = registry.sortComponentsByDependencies(components)
    const componentResults: Record<string, string> = {}

    for (const componentId of sortedComponents) {
      const component = registry.getComponent(componentId)
      if (!component) {
        if (!skipMissing) {
          throw new Error(`组件不存在: ${componentId}`)
        }
        continue
      }

      try {
        const originalTemplate = component.template
        if (templateOverrides[componentId]) {
          component.template = templateOverrides[componentId]
        }

        const result = await component.fn(context)

        if (originalTemplate !== undefined) {
          component.template = originalTemplate
        }

        if (result !== undefined) {
          componentResults[componentId] = result
        }
      } catch (error) {
        if (!skipMissing) {
          throw new Error(`组件 ${componentId} 构建失败: ${error}`)
        }
        console.warn(`组件 ${componentId} 构建失败，已跳过:`, error)
      }
    }

    return componentResults
  }

  /**
   * 组装提示词
   */
  private assemblePrompt(componentResults: Record<string, string>): string {
    // 简单的组装方式：按顺序连接所有组件
    const sections = Object.values(componentResults).filter(Boolean)

    if (sections.length === 0) {
      return ''
    }

    // 用双换行符分隔各个部分
    return sections.join('\n\n').trim()
  }
}

/**
 * 便捷函数：快速构建提示词
 */
export async function buildPrompt(
  context: ComponentContext,
  components: PromptComponent[],
  options: Omit<PromptBuildOptions, 'components'> = {}
): Promise<string> {
  const builder = new PromptBuilder()
  return builder.build(context, { ...options, components })
}

File: src-tauri/src/agent/prompt/builders/prompt_builder.rs
use std::collections::HashMap;

use serde_json::Value;

use crate::agent::config::PromptComponent;
use crate::agent::error::{AgentError, AgentResult};
use crate::agent::prompt::components::{
    registry::PromptComponentRegistry, types::ComponentContext,
};

/// Builder options for prompt building.
#[derive(Debug, Clone, Default)]
pub struct PromptBuildOptions {
    pub components: Vec<PromptComponent>,
    pub template_overrides: HashMap<PromptComponent, String>,
    pub additional_context: HashMap<String, Value>,
}

/// Core prompt builder coordinating component assembly.
pub struct PromptBuilder {
    registry: PromptComponentRegistry,
}

impl PromptBuilder {
    pub fn ensure_components_loaded(&mut self) {
        self.registry.ensure_loaded();
    }

    pub fn new() -> Self {
        Self {
            registry: PromptComponentRegistry::new(),
        }
    }

    pub async fn build(
        &mut self,
        mut context: ComponentContext,
        options: PromptBuildOptions,
    ) -> AgentResult<String> {
        context
            .additional_context
            .extend(options.additional_context.clone());

        let components = options.components.clone();
        let errors = self.registry.validate_dependencies(&components);
        if !errors.is_empty() {
            return Err(AgentError::Internal(format!(
                "Prompt component dependency errors: {}",
                errors.join(", ")
            )));
        }

        let sorted = self
            .registry
            .sort_by_dependencies(&components)
            .map_err(|e| AgentError::Internal(format!("Prompt component dependency cycle: {e}")))?;

        let mut rendered = HashMap::new();
        for component_id in sorted.iter() {
            let Some(def) = self.registry.get(component_id.clone()) else {
                return Err(AgentError::Internal(format!(
                    "Component not found: {:?}",
                    component_id
                )));
            };

            let template_override = options
                .template_overrides
                .get(component_id)
                .map(|s| s.as_str());

            if let Some(content) = def.render(&context, template_override).await? {
                rendered.insert(component_id.clone(), content);
            }
        }

        self.assemble_prompt(&sorted, &rendered)
    }

    fn assemble_prompt(
        &self,
        ordered_components: &[PromptComponent],
        rendered: &HashMap<PromptComponent, String>,
    ) -> AgentResult<String> {
        let mut sections = Vec::new();
        for component_id in ordered_components {
            if let Some(content) = rendered.get(component_id) {
                if !content.trim().is_empty() {
                    sections.push(content.clone());
                }
            }
        }
        Ok(sections.join("\n\n").trim().to_string())
    }
}

impl Default for PromptBuilder {
    fn default() -> Self {
        Self::new()
    }
}
